selectel/config.go

```diff
@@ -32,14 +46,74 @@ func (c *Config) Validate() error {
 			return err
 		}
 	}
+	if c.IdentityEndpoint == "" {
+		c.IdentityEndpoint = DefaultIdentityEndpoint
+	}
 
 	return nil
 }
 
+// Initialize Selectel resell client.
 func (c *Config) resellV2Client() *selvpcclient.ServiceClient {
 	return resellV2.NewV2ResellClientWithEndpoint(c.Token, c.Endpoint)
 }
 
+// Create Keystone token by Selectel token or Keystone credentials.
+func (c *Config) getToken(ctx context.Context, p string, r string) (string, error) {
+	if !c.hasKeystoneCredentials() {
+		return c.getTokenBySelectelToken(ctx, p, r)
+	}
+
+	return c.getTokenByCredentials(ctx, p, r)
+}
+
+// hasKeystoneCredentials determine Keystone credentials exists.
+func (c *Config) hasKeystoneCredentials() bool {
+	return c.User != "" && c.Password != "" && c.DomainName != ""
+}
+
+// Create Keystone token by Selectel token.
+func (c *Config) getTokenBySelectelToken(ctx context.Context, p string, r string) (string, error) {
+	tokenOpts := tokens.TokenOpts{
+		ProjectID: p,
+	}
+	resellV2Client := c.resellV2Client()
+	log.Print(msgCreate(objectToken, tokenOpts))
+
+	token, _, err := tokens.Create(ctx, resellV2Client, tokenOpts)
+	if err != nil {
+		return "", fmt.Errorf("create Keystone token by Selectel token: %w", err)
+	}
+
+	return token.ID, nil
+}
+
+// Create Keystone token by Keystone credentials.
+func (c *Config) getTokenByCredentials(ctx context.Context, p string, r string) (string, error) {
```

Comment:
the same

Reply:
Unused region variable removed from all functions - getToken, getTokenBySelectelToken, getTokenByCredentials, getDBaaSClient, getMKSClient, p renamed